src/rebase.js

```diff
@@ -116,13 +116,22 @@ module.exports = (function(){
   function _fetch(endpoint, options){
     _validateEndpoint(endpoint);
     optionValidators.context(options);
-    optionValidators.then(options);
     options.queries && optionValidators.query(options);
     var ref = firebase.database().ref(endpoint);
     ref = _addQueries(ref, options.queries);
-    ref.once('value', (snapshot) => {
+    return ref.once('value').then(snapshot => {
       var data = options.asArray === true ? _toArray(snapshot) : snapshot.val();
-      options.then.call(options.context, data);
+      if(options.then){
+        options.then.call(options.context, data);
+      }
+      return data;
+    }, err => {
+      //call onFailure callback if it exists otherwise return a rejected promise
+      if(options.onFailure && typeof options.onFailure === 'function'){
+        options.onFailure.call(options.context, err);
+      } else {
```

Comment:
@anotherpit 

### In re-base 2.1 

(using only callbacks)

Given an error ...

`fetch`
**with callback** _No Error at all, that was issue #124_ 
callback was required in `fetch`

`update`
**with callback** _Error is handled by callback, no unhandled exception_
**without optional callback**  _unhandled exception_

`post`
**with callback**  _Error is handled by callback, no unhandled exception_
**without optional callback**   _unhandled exception_

`push`
**with callback**  _Error is handled by callback, no unhandled exception_
**without optional callback**   _unhandled exception_

### In re-base 2.2

(using only callbacks)

Given an error ...

`fetch`
**with new onFailure callback** _Error is handled by callback, no unhandled exception_
**without new onFailure callback**  _unhandled exception_ (with  no `.catch` )

`update`
**with callback** _Error is handled by callback, no unhandled exception_
**without optional callback**  _unhandled exception_ (with  no `.catch` )

`post`
**with callback**  _Error is handled by callback, no unhandled exception_
**without optional callback**   _unhandled exception_ (with  no `.catch` )

`push`
**with callback**  _Error is handled by callback, no unhandled exception_
**without optional callback**   _unhandled exception_ (with  no `.catch` )

### With what you are proposing 
(returning the rejected Promise even if there is an `onFailure` callback passed to `fetch`)

`fetch`
**with new onFailure callback** _Error passed to callback **AND**  unhandled exception_
**without new onFailure callback**  _unhandled exception_ (with  no `.catch` )

As a user of re-base that IMO would be a bug. There is no way I can prevent an unhandled exception using `fetch` unless I use the Promise and add a `catch`.

The docs need to be updated to reflect that, since re-base 2.x, the optional callbacks are not really optional if you don't want an unhandled exception being thrown in your app. And since 2.2, they are only optional if you use the Promise. 

I understand your aversion to the implementation. It's a little odd. But that is also why I am glad re-base exists. So those couple lines can exist here,  so then in my app I can just

```javascript

fetch('data').then().catch()

```

**or**

```javascript
fetch('data', {
  then: onSuccess,
  onFailure: onFailure
});
```
And either approach will give me the expected behavior callbacks or Promises.